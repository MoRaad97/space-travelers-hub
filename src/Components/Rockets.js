import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { rocketsReducer } from '../Redux/Rockets/Rockets';
import './style/Rockets.css';

const Rockets = () => {
  // get rockets data from the store
  const rockets = useSelector((state) => state.rockets);

  const dispatch = useDispatch();

  const handle = (e) => {
    const { id } = e.target;
    dispatch(rocketsReducer.ReservationAction({ id }));
  };

  return (
    <section className="rockets">
      <div className="container">
        {rockets.map((rocket) => (
          <div className="rocket-card" key={rocket.rocket_id}>
            <img src={rocket.flickr_images[0]} alt={rocket.rocket_name} />
            <div>
              <h2>{rocket.rocket_name}</h2>
              <p>
                {rocket.reserved && (<span>Reserved</span>)}
                {rocket.description}
              </p>
              {!rocket.reserved && (
                <button
                  className="not-reserved"
                  type="button"
                  id={rocket.rocket_id}
                  onClick={handle}
                >
                  Reserve Rocket
                </button>
              )}
              {rocket.reserved && (
                <button
                  className="reserved"
                  type="button"
                  id={rocket.rocket_id}
                  onClick={handle}
                >
                  Cancel Reservation
                </button>
              )}
            </div>
          </div>
        ))}
      </div>
    </section>

  );
};

export default Rockets;
