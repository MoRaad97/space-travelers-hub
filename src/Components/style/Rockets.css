.rockets {
  font-family: Arial, Helvetica, sans-serif;
}

.rocket-card {
  display: flex;
  margin-bottom: 20px;
}

.rocket-card img {
  width: 267px;
  height: 200px;
  margin-right: 20px;
}

.rocket-card h2 {
  margin: 0;
  font-size: 22px;
  line-height: 1.5;
}

.rocket-card p {
  line-height: 1.5;
}

.rocket-card .not-reserved {
  padding: 10px;
  background-color: #0586ec;
  border: none;
  color: white;
  border-radius: 3px;
}

.rocket-card .reserved {
  padding: 10px;
  background-color: white;
  border: solid 1px;
  border-radius: 3px;
}

.rocket-card span {
  padding: 0 5px;
  background-color: #0586ec;
  color: white;
  border-radius: 3px;
  margin-right: 3px;
}
