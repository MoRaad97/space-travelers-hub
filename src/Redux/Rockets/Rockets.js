import axios from 'axios';
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
// Actions
const LOAD_ROCKETS = 'spacex-project-solo/rockets/LOAD_ROCKETS';

// Url
const baseUrl = 'https://api.spacexdata.com/v3/rockets';

// create ASYNC Thunks
export const loadRocketsThunk = createAsyncThunk(LOAD_ROCKETS, async () => {
  // fetch Rockets

  const response = await axios.get(`${baseUrl}`).catch((err) => {
    console.log('Error', err);
  });
  const res = response.data;
  return res;
});

const RocketsSlice = createSlice({
  name: 'spacex-project-solo/rockets',
  initialState: [],
  reducers: {
    ReservationAction(state, action) {
      const newState = state.map((object) => {
        if (object.rocket_id === action.payload.id) {
          return {
            ...object,
            reserved: !object.reserved,
          };
        }
        return {
          ...object,
        };
      });
      return newState;
    },
  },
  extraReducers: {
    [loadRocketsThunk.fulfilled]: (state, action) => {
      const newState = action.payload.map((object) => ({
        rocket_id: object.rocket_id,
        rocket_name: object.rocket_name,
        description: object.description,
        flickr_images: object.flickr_images,
        reserved: false,
      }));
      return newState;
    },
  },
});

export const rocketsReducer = RocketsSlice.actions;
export default RocketsSlice.reducer;
