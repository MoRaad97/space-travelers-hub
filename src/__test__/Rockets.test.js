import React from 'react';
import TestRenderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import store from '../Redux/configureStore';
import Rockets from '../Components/Rockets';

it('Rockets Component snapshot test', () => {
  const rocketsRender = TestRenderer
    .create(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    )
    .toJSON();
  expect(rocketsRender).toMatchSnapshot();
});
