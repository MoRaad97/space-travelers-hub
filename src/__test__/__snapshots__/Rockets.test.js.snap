// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rockets Component snapshot test 1`] = `
<section
  className="rockets"
>
  <div
    className="container"
  />
</section>
`;
